src/Typechecker/Monad.hs

```diff
@@ -164,84 +165,52 @@ hasType t1 t2 = if t1 <= t2 then return t2 else mismatch (Plain t1) (Plain t2)
 t :: Btype -> Typechecked Xtype
 t b = return (X b S.empty)
 
--- | Encoding the different type errors as types should let us do interesting things with them
-data TypeError = Mismatch {t1 :: Type,  t2 :: Type, srcPos2 :: (Expr SourcePos), srcPos :: SourcePos}                   -- ^ Couldn't match two types in an expression
```

Comment:
All of those duplicated `srcPos` were the initial inspiration for this new structure